spec/consensus/light-client.md

```diff
@@ -60,259 +61,368 @@ For the purpose of this lite client specification, we assume that the Tendermint
+  // validator set; otherwise false. Additional checks should be done in the implementation
+  // to ensure header is well formed.
+  func verify(h) bool {
+    vp_all := totalVotingPower(h.Header.V) // total sum of voting power of validators in h
+    return votingpower_in(signers(h.Commit),h.Header.V) > 2/3 * vp_all
+  }
 
-```go
-  func CheckSupport(h1,h2,trustlevel) bool {
-    if h1.Header.bfttime + tp < now { // Observation 1
-      return false // old header was once trusted but it is expired
-    }
-    vp_all := totalVotingPower(h1.Header.NextV)
-      // total sum of voting power of validators in h2
-
-    if h2.Header.height == h1.Header.height + 1 {
-      // specific check for adjacent headers; everything must be
-      // properly signed.
-      // also check that h2.Header.V == h1.Header.NextV
-      // Plus the following check that 2/3 of the voting power
-      // in h1 signed h2
-      return (votingpower_in(signers(h2.Commit),h1.Header.NextV) >
-              2/3 * vp_all)
-        // signing validators are more than two third in h1.
-    }
+  // Captures skipping condition. h1 and h2 have already passed basic validation
+  // (function `verify`).
+  // Returns nil in case h2 can be trusted based on h1, otherwise returns error.
+  // ErrHeaderExpired is used when h1 has expired with respect to lite client trusted period,
+  // ErrInvalidAdjacentHeaders when that adjacent headers are not consistent and
+  // ErrTooMuchChange when there is not enough intersection between validator sets to have
+  // skipping condition true.
+  func CheckSupport(h1,h2,trustThreshold) error {
+    assume h1.Header.Height < h2.header.Height and
+           h1.Header.bfttime < h2.Header.bfttime and
+           h2.Header.bfttime < now
+
+    if !isWithinTrustedPeriod(h1) return ErrHeaderNotWithinTrustedPeriod(h1)
+
+    // Although while executing the rest of CheckSupport function, h1 can expiry based
+    // on the lite client trusted period, this is not problem as lite client trusted
+    // period is smaller than trusted period of the header based on Tendermint Failure
+    // model, i.e., there is a significant time period (measure in days) during which
+    // validator set that has signed h1 can be trusted. Furthermore, CheckSupport function
+    // is not doing expensive operation (neither rpc nor signature verification), so it
+    // should execute fast.
+
+      // total sum of voting power of validators in h1.NextV
+      vp_all := totalVotingPower(h1.Header.NextV)
+
+      // check for adjacent headers
+      if (h2.Header.height == h1.Header.height + 1) {
+            if h1.Header.NextV == h2.Header.V 
+                return nil
+            return ErrInvalidAdjacentHeaders
+      }
+      // check for non-adjacent headers
+      if votingpower_in(signers(h2.Commit),h1.Header.NextV) > max(1/3,trustThreshold) * vp_all {
+        return nil
+      }
+      return ErrTooMuchChange
```

Comment:
Nit: This error was renamed in the go-code (at least in the current PR @melekes is working on). Would be cool if the reference implementation and the spec used the same names.